phylo/src/optimisers/topo_optimiser.rs

```diff
@@ -0,0 +1,122 @@
+
+        info!("Initial logl: {}.", initial_logl);
+        let mut curr_cost = initial_logl;
+        let mut prev_cost = f64::NEG_INFINITY;
+        let mut iterations = 0;
+
+        // No pruning on the root branch
+        let prune_locations: Vec<NodeIdx> = info
+            .tree
+            .preorder()
+            .iter()
+            .filter(|&n| n != &info.tree.root)
+            .cloned()
+            .collect();
+        while (curr_cost - prev_cost) > self.epsilon {
+            iterations += 1;
+            info!("Iteration: {}.", iterations);
+            prev_cost = curr_cost;
+            info!("Current logl: {}.", curr_cost);
+            for prune_branch in &prune_locations {
+                if info.tree.children(&info.tree.root).contains(prune_branch) {
+                    // due to topology change the current node may have become the direct child of root
+                    continue;
+                }
+                let regraft_locations = Self::find_regraft_options(prune_branch, &info);
+                let mut moves: Vec<(f64, Tree)> = Vec::with_capacity(regraft_locations.len());
+                for regraft_branch in &regraft_locations {
+                    let mut new_info = info.clone();
+                    new_info.tree = info.tree.rooted_spr(prune_branch, regraft_branch).unwrap();
+                    let mut logl = self.model.cost(&new_info, false);
+                    if logl <= curr_cost {
+                        // reoptimise branch length at the regraft location
+                        let o = BranchOptimiser::new(self.model, &new_info);
+                        let (blen_logl, blen) = o.optimise_branch(regraft_branch, &new_info)?;
+                        if blen_logl > logl {
+                            new_info.tree.set_blen(regraft_branch, blen);
+                            logl = blen_logl;
+                        }
+                    }
+                    moves.push((logl, new_info.tree.clone()));
+                }
+                let (best_logl, best_tree) = moves
+                    .into_iter()
+                    .max_by(|(a, _), (b, _)| a.partial_cmp(b).unwrap())
+                    .unwrap();
+                if best_logl > curr_cost {
+                    curr_cost = best_logl;
+                    info.tree = best_tree;
+                    debug!("Regrafted {} with new logl {}.", prune_branch, curr_cost);
+                } else {
+                    debug!(
+                        "No improvement regrafting {}, best logl {}.",
+                        prune_branch, best_logl
+                    );
+                }
+                // Optimise branch lengths on current tree to match PhyML
+                let o = BranchOptimiser::new(self.model, &info).run()?;
+                if o.final_logl > curr_cost {
+                    curr_cost = o.final_logl;
+                    info = o.i;
+                }
+                info.tree.clean(true);
```

Comment:
it isn't clear what this `true` is here

Reply:
This will be refactored in the next PR as per our discussion